sdk/python/kfp/dsl/type_utils.py

```diff
@@ -97,11 +108,12 @@ def get_parameter_type(
     AttributeError: if type_name is not a string type.
   """
   if type(param_type) == type:
-    if param_type not in (str, float, int):
-      raise TypeError('Got illegal parameter type. Currently only support: '
-                      'str, int and float. Got %s' % param_type)
-    param_type = param_type.__name__
-  return _PARAMETER_TYPES_MAPPING.get(param_type.lower())
+    type_name = param_type.__name__
+  elif isinstance(param_type, dict):
+    type_name = list(param_type.keys())[0]
+  else:
+    type_name = _data_passing._get_short_type_name(str(param_type))
```

Comment:
It's OK, but I'm not sure `component.yaml` files have `type: typing.List`, unless someone does that manually for some reason.
These generic types are expected to be converted to `JsonObject` or `JsonArray` when using `create_component_from_func`.
We usually try to use types that are not language-specific (python-specific).

Reply:
This would be needed for function-based component. 
`typing.List` is the case I had to make this change: https://github.com/kubeflow/pipelines/pull/5153